### PR TITLE
UWP: Add ability to filter logs and show event filter in Forms Puppet

### PR DESCRIPTION
**Pull request by guperrot**

### CODE REVIEW

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Channel/Channel.cs [hunk 1]
```diff
@@ -173,14 +216,21 @@ public void SetEnabled(bool enabled)
             }
         }
 
+        /// <summary>
+        /// Invalid device property cached information to refresh them next time we need it.
+        /// </summary>
```
Comment (achocron) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Channel/Channel.cs]: I don't really understand this comment. What is "them"?

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Channel/Channel.cs [hunk 2]
```diff
@@ -478,6 +532,9 @@ public Task ShutdownAsync()
             return Task.FromResult(default(object));
         }
 
+        /// <summary>
+        /// Free resources hold by this instance.
```
Comment (achocron) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Channel/Channel.cs]: hold -> held

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Channel/ChannelEventArgs.cs [hunk 3]
```diff
@@ -3,33 +3,98 @@
...
+        /// Set this property to true to request the log to be filtered out.
+        /// </summary>
+        public bool FilterRequested { get; set; }
+
+        /// <summary>
+        /// Init event with a log.
+        /// </summary>
+        /// <param name="log">log associated to this event.</param>
+        public FilteringLogEventArgs(Log log) : base(log) { }
+
```
Comment (achocron) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Channel/ChannelEventArgs.cs]: Extra line

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Channel/ChannelEventArgs.cs [hunk 4]
```diff
@@ -3,33 +3,98 @@
...
+    /// <summary>
+    /// Event called to possibly filter out a log before it is persisted and scheduled for sending.
+    /// </summary>
+    public class FilteringLogEventArgs : ChannelEventArgs
+    {
+        /// <summary>
+        /// Set this property to true to request the log to be filtered out.
+        /// </summary>
+        public bool FilterRequested { get; set; }
```
Comment (achocron) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Channel/ChannelEventArgs.cs]: Have you determined that the order of invocations won't affect whether this property is true or false when channel reads it?

File: Tests/Microsoft.AppCenter.Test.Windows/Channel/ChannelTest.cs [hunk 5]
```diff
@@ -149,6 +150,50 @@ public void EnqueueWhileDisabled()
...
         }
 
+        [TestMethod]
+        public void ChannelInvokesFilteringLogEvent()
+        {
+            for (var i = 0; i < MaxLogsPerBatch; ++i)
+            {
+                _channel.EnqueueAsync(new TestLog()).RunNotAsync();
+            }
+
```
Comment (achocron) [inline Tests/Microsoft.AppCenter.Test.Windows/Channel/ChannelTest.cs]: Extra line

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Channel/ChannelEventArgs.cs (hunk 4 above)
Comment (guperrot) [reply SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Channel/ChannelEventArgs.cs]: Order of invocations do affect the result but this is actually the official c# event pattern with bool return: https://docs.microsoft.com/en-us/dotnet/csharp/event-pattern#returning-values-from-event-subscribers

I even have a unit test to show filters canceling each other, this is by design.

File: Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet.UWP/EventFilter.cs [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+        #region instance
+
+        public override bool InstanceEnabled
+        {
+            get
+            {
+                return base.InstanceEnabled;
+            }
+
```
[the remark this answers is not in the document]
Comment (guperrot) [reply Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet.UWP/EventFilter.cs]: What is the issue? Here is how it looks like in VS:

capture

File: Tests/Microsoft.AppCenter.Test.Windows/Channel/ChannelTest.cs (hunk 5 above)
Comment (guperrot) [reply Tests/Microsoft.AppCenter.Test.Windows/Channel/ChannelTest.cs]: Fixed

File: Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet.UWP/EventFilter.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (achocron) [reply Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet.UWP/EventFilter.cs]: If that's intended i think there's another get/set pair where the line is missing. I just noticed the inconsistency

[the remark this answers is not in the document]
Comment (guperrot) [reply Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet.UWP/EventFilter.cs]: Where?

**Pull request merged**